### DISCUSSION TITLE
Sending Video Stream From the C++ Code

### DISCUSSION THREAD

**Discussion by LakshithaPW** [robotology/community#176] [Q&A] [answered]
Dear All,

I have a question about streaming a video to the iCub simulator screen.

I have successfully set up the iCub simulator and I can stream the example available in the README section:
```sh
$ yarpdev --device test_grabber --name /test/video --mode ball
$ yarp connect /test/video /icubSim/texture/screen
```
This works perfectly!

I would like to do the same from within my C++ code (e.g. project an image to the screen).
I could not find any help on the web for such a sample code. I tried to set up a `test_grabber` device like the following:
```cpp
PolyDriver texture_send;
Property p;
p.put("device", "test_grabber");
p.put("local", "/test/client"); 
p.put("remote", "/icubSim/texture/screen");
texture_send.open(p);
if (!texture_send.isValid()) {
    printf("Device not available.  Here are the known devices:\n");
    printf("%s", Drivers::factory().toString().c_str());
    exit(0);
}
// not sure...
```

The device can be opened up but could not find an interface.

Please let me know if you have any suggestions. 

Thanks in advance.
Cheers.

**Comment by pattacini** [comment]
Dear @robotology/ifi-padawans,

Can you please help @LakshithaPW to solve this problem?

**Comment by Nicogene** [comment, marked as answer]
Hi @LakshithaPW,

@giuliavezzani and I found some problems in your code, but we didn't test it, so after the changes we suggest you, if you have other problems let us know :smiley:

You have to put `mode` inside the `Property`, in order to run it in the same  way : `p.put("mode","ball")`.

Another problem is that you are adding the ports' names in fields `local` and `remote` that are not used by the `test_grabber` device. Our advice is to delete the two lines about remote and local and write `p.put("name","test/video")` and then `yarp connect /test/video /icubSim/texture/screen`.

@pattacini please correct us if we're wrong :sweat_smile:

Cheers,
Nicolo'

**Comment by LakshithaPW** [comment]
@Nicogene 
Thank you very much for the response.

I've changed the code to:

```cpp
PolyDriver texture_send;
Property p;
p.put("device","test_grabber");
p.put("name","test/video");
p.put("mode","ball");
texture_send.open(p);
if (!texture_send.isValid()) {
	printf("Device not available.  Here are the known devices:\n");
    printf("%s", Drivers::factory().toString().c_str());
    exit(0);
}
bool ok=Network::connect("/test/video","/icubSim/texture/screen");
if (!ok) {
    printf("Not Successful\n");
    exit(0);
}
```
There seems to be a small issue with the connection `Network::connect`. But it works well in the command prompt. In the connection step it seems nothing is happening. 

Please let me know if you notice any problem with this procedure.
Thanks in advance :)

**Comment by Nicogene** [comment]
You ought to:

```cpp
int main() {
    yarp::os::Network yarp;
    if (!yarp.checkNetwork()) {
        yarp::os::yError() << "YARP is not available!";
        return 1;
    }

    // the snippet above kicks in here
    // ...

    return 0;
}
```

However, it's way preferable to establish connections outside code, since we adhere to the [**publish-subscribe pattern**][1], which dictates that modules should remain as much **agnostic** as possible of the sources/sinks they receive/send data from/to.

[1]: https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern

**Comment by LakshithaPW** [comment]
Thanks @Nicogene and @pattacini I'll have a look.

**Comment by pattacini** [comment]
Closing for the time being then.
